moveit_ros/moveit_servo/include/moveit_servo/input_pipeline.hpp

```diff
@@ -0,0 +1,85 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/*      Title     : input_pipeline.hpp
+ *      Project   : moveit_servo
+ *      Created   : 11/06/2021
+ *      Author    : Tyler Weaver
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <rclcpp/rclcpp.hpp>
+#include <moveit_servo/input_pipeline.hpp>
+#include <moveit_servo/input_resampler.hpp>
+#include <moveit_servo/input_stale_command.hpp>
+#include <moveit_servo/input_subscriber.hpp>
+
+namespace moveit_servo
+{
+/**
+ * @brief      Builds the input pipeline based on the servo parameters
+ *
+ * @param[in]  node        The ros node
+ * @param[in]  parameters  The parameters
+ * @param[in]  halt        The halt function
+ * @param[in]  next        The next visitor connect the pipeline to
+ *
+ * @tparam     Visitor     The callable containing a visitor for the next step (normally ServoCalcs)
+ *
+ * @return     The input pipeline.
+ */
+template <typename Visitor>
+auto makeInputPipeline(const rclcpp::Node::SharedPtr& node, const ServoParameters& parameters,
+                       const std::function<void()>& halt, Visitor&& next)
+{
+  if (parameters.low_latency_mode)
+  {
+    return InputSubscriber{ node, parameters.cartesian_command_in_topic, parameters.joint_command_in_topic,
+                            std::move(next) };
+  }
+  else
+  {
+    auto stale_command_halt =
+        StaleCommandHalt{ node, rclcpp::Duration::from_seconds(parameters.incoming_command_timeout), halt,
+                          std::move(next) };
+    auto resampler = InputResampler{ node, parameters.publish_period, std::move(stale_command_halt) };
+    auto timestamp_now = TimestampNow{ node, std::move(resampler) };
+    return InputSubscriber{ node, parameters.cartesian_command_in_topic, parameters.joint_command_in_topic,
+                            std::move(timestamp_now) };
+  }
+}
```

Comment:
@griswaldbrooks here is my latest struggle.  What is the output type of this?  How might I store the output of this without resorting to dynamic memory?  Is this where I should use another variant?  I know what one of the output types is: `InputSubscriber<TimestampNow>`, but I do not know the output type of the other yet.  If possible I'd like to avoid putting that type in here because it is `ServoCalcs` which would make this much more difficult to test (it is not simple to mock, replace for testing).

For some context, much of the logic in this PR is currently buried in `ServoCalcs`.  I have not started reworking that to hook this up to it yet, so far I've just been building the various pieces and writing tests for them to verify the interfaces.

Another thing I'd like your critical review of is how I'm dealing with ownership of the Visitor objects.  As it is I'm giving ownership of the visitor object to the object that calls visit on it.  This simplifies dealing with the lifetime of each object.  I don't have to have some sort of container object for each step to manage their lifetime.  That is especially nice because I want to define which ones I construct at run time.

Reply:
Do you need to specify a return type? `auto` works in C++17. I'm pretty sure the copy is elided on the return here so you don't have to use `unique_ptr`, though I might leave the side-effect object out of it and build it more like
```
auto sub = InputSubscriber{node,
                                            parameters.cartesian_command_in_topic,
                                            parameters.joint_command_in_topic,
                                            makeInputPipeline(parameters, halter, []{})};
```
or some other factory for the `InputSubscriber` so you have more freedom to build and move pipeline objects.